exercises/diamond/test/Tests.hs

```diff
@@ -107,3 +134,40 @@ cases = [ Case { description = "Degenerate case with a single 'A' row"
                         "                         A                         "]
                }
         ]
+
+genNonAlphaChar :: Gen Char
+genNonAlphaChar = arbitraryASCIIChar `suchThat` (not . isLetter)
+
+genAlphaChar :: Gen Char
+genAlphaChar = arbitraryASCIIChar `suchThat` isUpper
```

Comment:
As I wrote in [this comment](https://github.com/exercism/haskell/pull/843#discussion_r321981808), ``gen `suchThat` predicate`` has a drawback when `predicate` accepts only a few of the values that `gen` generates, since at best it will be inefficient at generating, and at worst it will give up. You can see this in [the source code for `suchThat`](http://hackage.haskell.org/package/QuickCheck-2.13.2/docs/src/Test.QuickCheck.Gen.html#suchThat):

```haskell
-- | Generates a value that satisfies a predicate.
suchThat :: Gen a -> (a -> Bool) -> Gen a
gen `suchThat` p =
  do mx <- gen `suchThatMaybe` p
     case mx of
       Just x  -> return x
       Nothing -> sized (\n -> resize (n+1) (gen `suchThat` p))

-- | Tries to generate a value that satisfies a predicate.
-- If it fails to do so after enough attempts, returns @Nothing@.
suchThatMaybe :: Gen a -> (a -> Bool) -> Gen (Maybe a)
gen `suchThatMaybe` p = sized (\n -> try n (2*n))
 where
  try m n
    | m > n = return Nothing
    | otherwise = do
        x <- resize m gen
        if p x then return (Just x) else try (m+1) n
```

Here is a micro-benchmark using [Criterion](http://www.serpentine.com/criterion/tutorial.html):

```haskell
module Main where

import Test.QuickCheck
import Criterion
import Criterion.Main
import Data.Char (isLetter, isUpper)
import Data.List ((\\))

main :: IO ()
main = defaultMain
  [ bgroup "genUpper"
      [ bench "suchThat"        $ genUpperSuchThat       `nfQC` isUpper
      , bench "elements"        $ genUpperElements       `nfQC` isUpper
      , bench "elements/filter" $ genUpperElementsFilter `nfQC` isUpper
      ]
  , bgroup "genNonAlpha"
      [ bench "suchThat"        $ genNonLetterSuchThat       `nfQC` (not . isUpper)
      , bench "elements"        $ genNonLetterElements       `nfQC` (not . isUpper)
      , bench "elements/filter" $ genNonLetterElementsFilter `nfQC` (not . isUpper)
      ]
  ]

nfQC :: Gen Char -> (Char -> Bool) -> Benchmarkable
nfQC gen = nfIO . quickCheckWith args . forAll gen
  where
    args = stdArgs { chatty = False }

genUpperSuchThat, genUpperElements, genUpperElementsFilter :: Gen Char
genUpperSuchThat = arbitraryASCIIChar `suchThat` isUpper
genUpperElements = elements ['A'..'Z']
genUpperElementsFilter = elements . filter isUpper $ ['\0'..'\127']

genNonLetterSuchThat, genNonLetterElements, genNonLetterElementsFilter :: Gen Char
genNonLetterSuchThat = arbitraryASCIIChar `suchThat` (not . isLetter)
genNonLetterElements = elements (['\0'..'\127'] \\ (['A'..'Z'] ++ ['a'..'z']))
genNonLetterElementsFilter = elements . filter (not . isLetter) $ ['\0'..'\127']
```

and compiling and running it gives:

```
$ stack ghc -- -O2 AsciiCharBenchmark.hs --output gen.html
[1 of 1] Compiling Main             ( AsciiCharBenchmark.hs, AsciiCharBenchmark.o )
Linking AsciiCharBenchmark ...

$ ./AsciiCharBenchmark 
benchmarking genUpper/suchThat
time                 213.5 μs   (212.0 μs .. 216.1 μs)
                     0.999 R²   (0.998 R² .. 0.999 R²)
mean                 218.1 μs   (215.8 μs .. 221.8 μs)
std dev              9.044 μs   (6.284 μs .. 13.24 μs)
variance introduced by outliers: 39% (moderately inflated)

benchmarking genUpper/elements
time                 89.69 μs   (88.31 μs .. 91.90 μs)
                     0.994 R²   (0.987 R² .. 1.000 R²)
mean                 89.43 μs   (88.29 μs .. 92.76 μs)
std dev              5.595 μs   (1.949 μs .. 10.41 μs)
variance introduced by outliers: 63% (severely inflated)

benchmarking genUpper/elements/filter
time                 89.38 μs   (89.07 μs .. 89.73 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 89.09 μs   (88.92 μs .. 89.31 μs)
std dev              609.7 ns   (441.7 ns .. 825.1 ns)

benchmarking genNonAlpha/suchThat
time                 124.2 μs   (121.9 μs .. 127.2 μs)
                     0.998 R²   (0.997 R² .. 1.000 R²)
mean                 122.4 μs   (121.8 μs .. 123.7 μs)
std dev              2.859 μs   (1.572 μs .. 5.033 μs)
variance introduced by outliers: 19% (moderately inflated)

benchmarking genNonAlpha/elements
time                 95.06 μs   (94.96 μs .. 95.17 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 95.06 μs   (94.96 μs .. 95.21 μs)
std dev              404.3 ns   (251.1 ns .. 626.1 ns)

benchmarking genNonAlpha/elements/filter
time                 95.00 μs   (94.85 μs .. 95.15 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 95.22 μs   (95.07 μs .. 95.56 μs)
std dev              707.7 ns   (442.0 ns .. 1.240 μs)
```

With HTML output [here](https://simonshine.dk/gen.html) (the colors seem a bit off).

As you can see, `suchThat` is less efficient when only 26/128 generated elements are accepted by its right-hand side predicate, as when (128 - 2*26)/128 = 76/128 generated elements are accepted. Generating from a positive set is always faster. So while it doesn't really matter here, the point I wanted to make was that you should be careful with `suchThat`.

I said something wrong to you yesterday, being that `suchThat` would eventually make the test give up. But as evidenced from:

```haskell
     case mx of
       Just x  -> return x
       Nothing -> sized (\n -> resize (n+1) (gen `suchThat` p))
```

it actually keeps growing the generator's size parameter.

Reply:
Considering this, I decided that the `genNonAlphaChar` is more declarative using `suchThat`. However, `genAlphaChar` is cleaner using `elements`. If we consider the performance data `elements` is more compelling for the latter case too.